include/envoy/config/discovery_service_base.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/api/v2/auth/cert.pb.h"
+#include "envoy/api/v2/cluster.pb.h"
+#include "envoy/api/v2/endpoint.pb.h"
+#include "envoy/api/v2/listener.pb.h"
+#include "envoy/api/v2/rds.pb.h"
+#include "envoy/api/v2/route/route_components.pb.h"
+#include "envoy/api/v2/scoped_route.pb.h"
+#include "envoy/config/cluster/v3/cluster.pb.h"
+#include "envoy/config/core/v3/config_source.pb.h"
+#include "envoy/config/discovery_service_base.h"
+#include "envoy/config/endpoint/v3/endpoint.pb.h"
+#include "envoy/config/listener/v3/listener.pb.h"
+#include "envoy/config/route/v3/route.pb.h"
+#include "envoy/config/route/v3/route_components.pb.h"
+#include "envoy/config/route/v3/scoped_route.pb.h"
+#include "envoy/config/subscription.h"
+#include "envoy/extensions/transport_sockets/tls/v3/cert.pb.h"
+#include "envoy/service/discovery/v2/rtds.pb.h"
+#include "envoy/service/runtime/v3/rtds.pb.h"
+
+#define RESOURCE_NAME_GETTER_REGISTRATION(PREV_RESOURCE, CURRENT_RESOURCE)                         \
```

Comment:
LGTM 
Parametered static member methods reminds me of `Policy-Based Class Design` but it doesn't make big difference here

Reply:
@lambdai I think that It is more readable and clear to be aggregated because all of discovery service subscription has `loadTypeUrl`. It does't make big difference but clarity.